File: aws/ses/typescript/src/services/templates.ts
/**
 * Template Service
 *
 * Service for managing email templates in AWS SES v2.
 *
 * @module services/templates
 */

import { BaseService } from './base.js';
import type {
  CreateEmailTemplateRequest,
  UpdateEmailTemplateRequest,
  DeleteEmailTemplateRequest,
  GetEmailTemplateRequest,
  ListEmailTemplatesRequest,
  TestRenderEmailTemplateRequest,
} from '../types/requests.js';
import type {
  CreateEmailTemplateResponse,
  UpdateEmailTemplateResponse,
  DeleteEmailTemplateResponse,
  GetEmailTemplateResponse,
  ListEmailTemplatesResponse,
  TestRenderEmailTemplateResponse,
} from '../types/responses.js';

/**
 * Template service for managing email templates.
 *
 * Provides methods for:
 * - Creating email templates
 * - Updating email templates
 * - Deleting email templates
 * - Retrieving template details
 * - Listing all templates
 * - Testing template rendering
 *
 * @example
 * ```typescript
 * const templateService = new TemplateService(client);
 *
 * // Create a template
 * await templateService.createEmailTemplate({
 *   templateName: 'welcome-email',
 *   templateContent: {
 *     subject: 'Welcome {{name}}!',
 *     html: '<h1>Hello {{name}}</h1>',
 *     text: 'Hello {{name}}'
 *   }
 * });
 *
 * // Test rendering
 * const result = await templateService.testRenderEmailTemplate(
 *   'welcome-email',
 *   JSON.stringify({ name: 'Alice' })
 * );
 * console.log(result.renderedEmailTemplate);
 * ```
 */
export class TemplateService extends BaseService {
  /**
   * Create an email template.
   *
   * Creates a new email template that can be used for sending personalized emails.
   * Templates support variable substitution using {{variableName}} syntax.
   *
   * @param request - Template creation request
   * @returns Promise resolving when template is created
   *
   * @throws {TemplateError} If template already exists
   * @throws {ValidationError} If template content is invalid
   *
   * @example
   * ```typescript
   * await templateService.createEmailTemplate({
   *   templateName: 'newsletter',
   *   templateContent: {
   *     subject: '{{monthName}} Newsletter',
   *     html: `
   *       <html>
   *         <body>
   *           <h1>Hello {{firstName}}!</h1>
   *           <p>Here's your {{monthName}} newsletter.</p>
   *           <a href="{{unsubscribeUrl}}">Unsubscribe</a>
   *         </body>
   *       </html>
   *     `,
   *     text: 'Hello {{firstName}}! Newsletter for {{monthName}}.'
   *   }
   * });
   * ```
   */
  async createEmailTemplate(request: CreateEmailTemplateRequest): Promise<CreateEmailTemplateResponse> {
    return this.post<CreateEmailTemplateResponse>('/v2/email/templates', request);
  }

  /**
   * Delete an email template.
   *
   * Permanently deletes an email template. This operation cannot be undone.
   *
   * @param templateName - Name of the template to delete
   * @returns Promise resolving when template is deleted
   *
   * @throws {TemplateError} If template does not exist
   *
   * @example
   * ```typescript
   * await templateService.deleteEmailTemplate('old-template');
   * ```
   */
  async deleteEmailTemplate(templateName: string): Promise<void> {
    return this.delete(`/v2/email/templates/${encodeURIComponent(templateName)}`);
  }

  /**
   * Get email template details.
   *
   * Retrieves the content and metadata for an email template.
   *
   * @param templateName - Name of the template to retrieve
   * @returns Promise resolving to template details
   *
   * @throws {TemplateError} If template does not exist
   *
   * @example
   * ```typescript
   * const template = await templateService.getEmailTemplate('welcome-email');
   * console.log('Subject:', template.templateContent?.subject);
   * console.log('HTML:', template.templateContent?.html);
   * console.log('Text:', template.templateContent?.text);
   * ```
   */
  async getEmailTemplate(templateName: string): Promise<GetEmailTemplateResponse> {
    return this.get<GetEmailTemplateResponse>(
      `/v2/email/templates/${encodeURIComponent(templateName)}`
    );
  }

  /**
   * List email templates.
   *
   * Returns a list of all email templates in your account.
   * Results are paginated if there are many templates.
   *
   * @param nextToken - Token for pagination (from previous response)
   * @param pageSize - Maximum number of templates to return (1-100)
   * @returns Promise resolving to list of templates
   *
   * @example
   * ```typescript
   * // Get first page
   * let response = await templateService.listEmailTemplates();
   * console.log('Templates:', response.templatesMetadata);
   *
   * // Get next page if available
   * if (response.nextToken) {
   *   response = await templateService.listEmailTemplates(response.nextToken);
   * }
   * ```
   *
   * @example Iterate through all templates
   * ```typescript
   * let nextToken: string | undefined;
   * const allTemplates = [];
   *
   * do {
   *   const response = await templateService.listEmailTemplates(nextToken, 50);
   *   if (response.templatesMetadata) {
   *     allTemplates.push(...response.templatesMetadata);
   *   }
   *   nextToken = response.nextToken;
   * } while (nextToken);
   *
   * console.log(`Total templates: ${allTemplates.length}`);
   * ```
   */
  async listEmailTemplates(
    nextToken?: string,
    pageSize?: number
  ): Promise<ListEmailTemplatesResponse> {
    const query = this.buildQuery({
      NextToken: nextToken,
      PageSize: pageSize,
    });

    return this.get<ListEmailTemplatesResponse>('/v2/email/templates', query);
  }

  /**
   * Update an email template.
   *
   * Updates the content of an existing email template.
   * All template content fields are optional - only provided fields will be updated.
   *
   * @param request - Template update request
   * @returns Promise resolving when template is updated
   *
   * @throws {TemplateError} If template does not exist
   * @throws {ValidationError} If template content is invalid
   *
   * @example
   * ```typescript
   * // Update only the subject line
   * await templateService.updateEmailTemplate({
   *   templateName: 'welcome-email',
   *   templateContent: {
   *     subject: 'Welcome to {{companyName}}!'
   *   }
   * });
   *
   * // Update all content
   * await templateService.updateEmailTemplate({
   *   templateName: 'newsletter',
   *   templateContent: {
   *     subject: '{{monthName}} Update',
   *     html: '<h1>{{monthName}} Newsletter</h1><p>{{content}}</p>',
   *     text: '{{monthName}} Newsletter: {{content}}'
   *   }
   * });
   * ```
   */
  async updateEmailTemplate(request: UpdateEmailTemplateRequest): Promise<UpdateEmailTemplateResponse> {
    return this.put<UpdateEmailTemplateResponse>(
      `/v2/email/templates/${encodeURIComponent(request.templateName)}`,
      { templateContent: request.templateContent }
    );
  }

  /**
   * Test render an email template.
   *
   * Renders an email template with the provided data to preview how it will look.
   * This is useful for testing templates before sending them to recipients.
   *
   * @param templateName - Name of the template to render
   * @param templateData - JSON string containing template variable values
   * @returns Promise resolving to rendered email content
   *
   * @throws {TemplateError} If template does not exist or is invalid
   * @throws {ValidationError} If template data is invalid JSON
   *
   * @example
   * ```typescript
   * const result = await templateService.testRenderEmailTemplate(
   *   'welcome-email',
   *   JSON.stringify({
   *     name: 'Alice',
   *     companyName: 'Acme Corp',
   *     loginUrl: 'https://app.example.com/login'
   *   })
   * );
   *
   * console.log('Subject:', result.subject);
   * console.log('HTML:', result.htmlPart);
   * console.log('Text:', result.textPart);
   * ```
   *
   * @example Testing with different data sets
   * ```typescript
   * const testCases = [
   *   { name: 'Short Name', companyName: 'A' },
   *   { name: 'Long Name With Many Words', companyName: 'Very Long Company Name Inc.' },
   *   { name: 'Special &lt;chars&gt;', companyName: 'Test & Co.' }
   * ];
   *
   * for (const data of testCases) {
   *   const result = await templateService.testRenderEmailTemplate(
   *     'welcome-email',
   *     JSON.stringify(data)
   *   );
   *   console.log('Test case:', data.name);
   *   console.log('Rendered subject:', result.subject);
   * }
   * ```
   */
  async testRenderEmailTemplate(
    templateName: string,
    templateData: string
  ): Promise<TestRenderEmailTemplateResponse> {
    return this.post<TestRenderEmailTemplateResponse>(
      `/v2/email/templates/${encodeURIComponent(templateName)}/render`,
      { templateData }
    );
  }
}

File: aws/ses/rust/src/services/templates.rs
//! Email template operations for SES v2.
//!
//! This module provides methods for managing email templates, including
//! creating, updating, deleting, listing, and testing template rendering.

use std::sync::Arc;
use serde::{Deserialize, Serialize};

use crate::error::{SesError, SesResult};
use crate::http::{HttpClient, SesRequest, HttpMethod};
use super::SesService;

/// Service for email template operations.
///
/// This service provides methods for:
/// - Creating email templates
/// - Updating existing templates
/// - Deleting templates
/// - Listing all templates
/// - Getting template details
/// - Testing template rendering with sample data
pub struct TemplateService {
    http_client: Arc<dyn HttpClient>,
}

impl TemplateService {
    /// Create a new template service.
    pub fn new(http_client: impl HttpClient + 'static) -> Self {
        Self {
            http_client: Arc::new(http_client),
        }
    }

    /// Create a new email template.
    ///
    /// # Arguments
    ///
    /// * `template_name` - Unique name for the template
    /// * `subject` - Email subject (may contain {{variables}})
    /// * `html` - Optional HTML body
    /// * `text` - Optional text body
    ///
    /// # Errors
    ///
    /// Returns an error if a template with the same name already exists.
    pub async fn create_email_template(
        &self,
        template_name: &str,
        subject: Option<String>,
        html: Option<String>,
        text: Option<String>,
    ) -> SesResult<CreateEmailTemplateResponse> {
        let request = CreateEmailTemplateRequest {
            template_name: template_name.to_string(),
            template_content: TemplateContent { subject, html, text },
        };

        let body = serde_json::to_vec(&request)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize CreateEmailTemplate request: {}", e),
            })?;

        let ses_request = SesRequest::new(HttpMethod::Post, "/v2/email/templates")
            .with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize CreateEmailTemplate response: {}", e),
            })
    }

    /// Delete an email template.
    ///
    /// # Arguments
    ///
    /// * `template_name` - Name of the template to delete
    pub async fn delete_email_template(
        &self,
        template_name: &str,
    ) -> SesResult<DeleteEmailTemplateResponse> {
        let path = format!("/v2/email/templates/{}", template_name);
        let ses_request = SesRequest::new(HttpMethod::Delete, &path);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize DeleteEmailTemplate response: {}", e),
            })
    }

    /// Get an email template.
    ///
    /// # Arguments
    ///
    /// * `template_name` - Name of the template to retrieve
    ///
    /// # Returns
    ///
    /// The template content including subject, HTML, and text parts.
    pub async fn get_email_template(
        &self,
        template_name: &str,
    ) -> SesResult<GetEmailTemplateResponse> {
        let path = format!("/v2/email/templates/{}", template_name);
        let ses_request = SesRequest::new(HttpMethod::Get, &path);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize GetEmailTemplate response: {}", e),
            })
    }

    /// List all email templates.
    ///
    /// # Arguments
    ///
    /// * `next_token` - Token for pagination (from previous response)
    /// * `page_size` - Maximum number of results to return (1-100)
    ///
    /// # Returns
    ///
    /// A list of template metadata and an optional next token for pagination.
    pub async fn list_email_templates(
        &self,
        next_token: Option<String>,
        page_size: Option<i32>,
    ) -> SesResult<ListEmailTemplatesResponse> {
        let mut path = "/v2/email/templates".to_string();
        let mut query_params = Vec::new();

        if let Some(token) = next_token {
            query_params.push(format!("NextToken={}", token));
        }
        if let Some(size) = page_size {
            query_params.push(format!("PageSize={}", size));
        }

        if !query_params.is_empty() {
            path.push_str("?");
            path.push_str(&query_params.join("&"));
        }

        let ses_request = SesRequest::new(HttpMethod::Get, &path);
        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize ListEmailTemplates response: {}", e),
            })
    }

    /// Update an existing email template.
    ///
    /// # Arguments
    ///
    /// * `template_name` - Name of the template to update
    /// * `subject` - Updated subject line
    /// * `html` - Updated HTML body
    /// * `text` - Updated text body
    pub async fn update_email_template(
        &self,
        template_name: &str,
        subject: Option<String>,
        html: Option<String>,
        text: Option<String>,
    ) -> SesResult<UpdateEmailTemplateResponse> {
        let request = UpdateEmailTemplateRequest {
            template_content: TemplateContent { subject, html, text },
        };

        let body = serde_json::to_vec(&request)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize UpdateEmailTemplate request: {}", e),
            })?;

        let path = format!("/v2/email/templates/{}", template_name);
        let ses_request = SesRequest::new(HttpMethod::Put, &path).with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize UpdateEmailTemplate response: {}", e),
            })
    }

    /// Test rendering an email template with sample data.
    ///
    /// # Arguments
    ///
    /// * `template_name` - Name of the template to test
    /// * `template_data` - JSON string with template variables
    ///
    /// # Returns
    ///
    /// The rendered subject, HTML, and text with variables replaced.
    pub async fn test_render_email_template(
        &self,
        template_name: &str,
        template_data: &str,
    ) -> SesResult<TestRenderEmailTemplateResponse> {
        let request = TestRenderEmailTemplateRequest {
            template_data: template_data.to_string(),
        };

        let body = serde_json::to_vec(&request)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to serialize TestRenderEmailTemplate request: {}", e),
            })?;

        let path = format!("/v2/email/templates/{}/render", template_name);
        let ses_request = SesRequest::new(HttpMethod::Post, &path).with_body(body);

        let response = self.http_client.send_request(ses_request).await?;

        serde_json::from_slice(&response.body)
            .map_err(|e| SesError::Serialization {
                message: format!("Failed to deserialize TestRenderEmailTemplate response: {}", e),
            })
    }
}

impl SesService for TemplateService {
    fn http_client(&self) -> &Arc<dyn HttpClient> {
        &self.http_client
    }
}

// Request types

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct CreateEmailTemplateRequest {
    template_name: String,
    template_content: TemplateContent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct TemplateContent {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub html: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct UpdateEmailTemplateRequest {
    template_content: TemplateContent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
struct TestRenderEmailTemplateRequest {
    template_data: String,
}

// Response types

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct CreateEmailTemplateResponse {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct DeleteEmailTemplateResponse {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct GetEmailTemplateResponse {
    pub template_name: String,
    pub template_content: TemplateContent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct ListEmailTemplatesResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub templates_metadata: Option<Vec<EmailTemplateMetadata>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_token: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct EmailTemplateMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub template_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_timestamp: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct UpdateEmailTemplateResponse {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct TestRenderEmailTemplateResponse {
    pub rendered_subject: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rendered_html_part: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rendered_text_part: Option<String>,
}
